### DISCUSSION TITLE
RuntimeError: Unexpected ASGI message 'websocket.send', after sending 'websocket.close'

### DISCUSSION THREAD

**Discussion by nandu1336** [fastapi/fastapi#8919] [Questions] [answered]
Hey !! 

- I am trying to build a chat server with two websocket endpoints. one for the room owners to create_room and second for others who wants to join the room. 
- I have 'Room' and 'RoomManager' classes. Room has list of members, and other room info. where as 'RoomManager' helps in creating rooms and accepting people into the room. 
- when a user tries to enter the room with room_id i am sending a request to the room_owner whether to add the member to the room. when user accepts I am adding the member to the list of memberes in the room. 
- when a new user tries to enter a room and the owner accept it (by sending 'yes')  I wanted to broadcast "**xxxx has entered the room**". when I try to broadcast it shows this error "**RuntimeError: Unexpected ASGI message 'websocket.send', after sending 'websocket.close'** ".  [this is the link](https://github.com/nandu1336/spare-chats/tree/master/src) to repo. Please help me resolve the issue. I heard of sending ping pong frames but did not find docs to implement it.

files explanation in repo:

- app.py : contains http and websocket endpoints
- room_manager.py : has Room and RoomManager classes.
-  connection_manger.py: default connection manager provided by fastapi docs. used to store temporary connections before user is part of any room.

**Below is the cmd logs:**

> H:\PYTHON\chats\src>uvicorn app:app --reload
> INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
> INFO:     Started reloader process [12408] using watchgod
> INFO:     Started server process [6908]
> INFO:     Waiting for application startup.
> INFO:     Application startup complete.
> INFO:     ('127.0.0.1', 61529) - "WebSocket /create_room/{"room_name":"new room","owner_name":"owner","owner_id":"472a97ab-c5a8-456c-97e0-487346be2ac9"}" [accepted]
> room_created || room_code: 9A9Ebf
> connecet in manager:
> INFO:     ('127.0.0.1', 61334) - "WebSocket /join_room/{"room_code":"9A9Ebf","username":"member 2","user_id":"6c47a7bf-325e-4ab8-ba56-9afa3a6c086c"}" [accepted]
> self.active_members.keys() dict_keys(['6c47a7bf-325e-4ab8-ba56-9afa3a6c086c'])
> self.active_rooms.keys():: dict_keys(['9A9Ebf'])
> self.active_rooms[room_code]:: <room_manager.Room object at 0x000001ED9E6D3B70>
> control inside Room.join_room:: {'room_code': '9A9Ebf', 'username': 'member 2', 'user_id': '6c47a7bf-325e-4ab8-ba56-9afa3a6c086c', 'ws': <starlette.websockets.WebSocket object at 0x000001ED9E709668>}
> trying to send_personal_message to owner
> user_id: 6c47a7bf-325e-4ab8-ba56-9afa3a6c086c
> user in 105 : {'room_code': '9A9Ebf', 'username': 'member 2', 'user_id': '6c47a7bf-325e-4ab8-ba56-9afa3a6c086c', 'ws': <starlette.websockets.WebSocket object at 0x000001ED9E709668>}
> userDetial sin 89: {'room_code': '9A9Ebf', 'username': 'member 2', 'user_id': '6c47a7bf-325e-4ab8-ba56-9afa3a6c086c', 'ws': <starlette.websockets.WebSocket object at 0x000001ED9E709668>}
> ERROR:    Exception in ASGI application
> Traceback (most recent call last):
>   File "c:\users\nandu\appdata\local\programs\python\python37\lib\site-packages\uvicorn\protocols\websockets\websockets_impl.py", line 171, in run_asgi
>     result = await self.app(self.scope, self.asgi_receive, self.asgi_send)
>   File "c:\users\nandu\appdata\local\programs\python\python37\lib\site-packages\uvicorn\middleware\proxy_headers.py", line 59, in __call__
>     return await self.app(scope, receive, send)
>   File "c:\users\nandu\appdata\local\programs\python\python37\lib\site-packages\fastapi\applications.py", line 199, in __call__
>     await super().__call__(scope, receive, send)
>   File "c:\users\nandu\appdata\local\programs\python\python37\lib\site-packages\starlette\applications.py", line 112, in __call__
>     await self.middleware_stack(scope, receive, send)
>   File "c:\users\nandu\appdata\local\programs\python\python37\lib\site-packages\starlette\middleware\errors.py", line 146, in __call__
>     await self.app(scope, receive, send)
>   File "c:\users\nandu\appdata\local\programs\python\python37\lib\site-packages\starlette\middleware\cors.py", line 70, in __call__
>     await self.app(scope, receive, send)
>   File "c:\users\nandu\appdata\local\programs\python\python37\lib\site-packages\starlette\exceptions.py", line 58, in __call__
>     await self.app(scope, receive, send)
>   File "c:\users\nandu\appdata\local\programs\python\python37\lib\site-packages\starlette\routing.py", line 580, in __call__
>     await route.handle(scope, receive, send)
>   File "c:\users\nandu\appdata\local\programs\python\python37\lib\site-packages\starlette\routing.py", line 297, in handle
>     await self.app(scope, receive, send)
>   File "c:\users\nandu\appdata\local\programs\python\python37\lib\site-packages\starlette\routing.py", line 68, in app
>     await func(session)
>   File "c:\users\nandu\appdata\local\programs\python\python37\lib\site-packages\fastapi\routing.py", line 260, in app
>     await dependant.call(**values)
>   File ".\app.py", line 89, in create_room
>     await room.listen_for_joiners()
>   File ".\room_manager.py", line 112, in listen_for_joiners
>     await self.add_user_to_room(user,user['ws'])
>   File ".\room_manager.py", line 93, in add_user_to_room
>     await self.broadcast_message(f"{user_details['username']} Has Entered The Chat.")
>   File ".\room_manager.py", line 79, in broadcast_message
>     await member_socket.send_text(message)
>   File "c:\users\nandu\appdata\local\programs\python\python37\lib\site-packages\starlette\websockets.py", line 128, in send_text
>     await self.send({"type": "websocket.send", "text": data})
>   File "c:\users\nandu\appdata\local\programs\python\python37\lib\site-packages\starlette\websockets.py", line 68, in send
>     await self._send(message)
>   File "c:\users\nandu\appdata\local\programs\python\python37\lib\site-packages\uvicorn\protocols\websockets\websockets_impl.py", line 249, in asgi_send
>     raise RuntimeError(msg % message_type)
> RuntimeError: Unexpected ASGI message 'websocket.send', after sending 'websocket.close'.
>

**Comment by alfiephillips** [comment, marked as answer]
There is a similar issue here that was solved. https://github.com/tiangolo/fastapi/issues/2370
